### PR TITLE
[Fix-14503][Worker] Fix the problem of subprocess cannot be killed

### LINKED ISSUE #14503: [Bug] [Worker] Cannot kill the subprocess
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

worker cannot kill the subprocess

### What you expected to happen

subprocess could be killed

### How to reproduce

* create a shell task with `sleep 300`
* kill the task
* `sleep 300` is still running

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by rickchengx**
## Purpose of the pull request

* close: #14503 

## Brief change log

* `doKill()` must be running before `cancelApplication()` so that `pstree` could get all the sub process id and kill them

## Verify this pull request

截屏2023-07-11 14 46 38

### CODE REVIEW

**Comment by rickchengx** [comment]
Hi, @Radeity could you please help review this

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/WorkerTaskKillProcessor.java [hunk 1]
```diff
@@ -98,6 +96,7 @@ public void process(Channel channel, Message message) {
             }
 
             boolean result = doKill(taskExecutionContext);
+            this.cancelApplication(taskInstanceId);
```
Comment (ruanwenjun) [inline dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/WorkerTaskKillProcessor.java]: I hope we can remove the `doKill` method, the kill operation should be implemented by task plugin.

**Comment by ruanwenjun** [comment]
This is strange, when we kill the shell task, we will use `process.destroy();` to cancel the process.

I test this case, and find the process has been killed, but throw exception since the kill operation is not permitted.

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/WorkerTaskKillProcessor.java (hunk 1 above)
Comment (rickchengx) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/WorkerTaskKillProcessor.java]: @ruanwenjun  Do u mean that all the system process (include subprocess) and the remote process (like yarn) should be killed in `cancelApplication` in each task plugin ?
* but the `process.destroy()` cannot kill the subprocess. like `sleep 300` in the shell task. the `sleep 300`
 is still running after we stop the workflow instance.
截屏2023-07-11 14 31 07

* And also the remote process now is killed in `doKill()`, not in each task plugin.

Comment (ruanwenjun) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/WorkerTaskKillProcessor.java]: I cannot reproduce this on my mac.

Comment (rickchengx) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/processor/WorkerTaskKillProcessor.java]: > I cannot reproduce this on my mac.

I think in some cases, `process.destroy` cannot kill the subprocess.
截屏2023-07-13 10 12 39
https://stackoverflow.com/questions/5879791/how-to-kill-subprocesses-of-a-java-process

**Comment by rickchengx** [comment]
> I cannot reproduce this on my mac.

### How to reproduce
> create a shell task with sleep 300
kill the task
sleep 300 is still running


Hi, @zhongjiajie @Radeity @SbloodyS  Could you guys please take a time to see if this bug can be reproduced?

On my mac, the task is killed at 14:48
image

but `sleep 300` is still running at 14:49
截屏2023-07-13 14 50 13

**Comment by Radeity** [comment]
> > I cannot reproduce this on my mac.
> 
> ### How to reproduce
> > create a shell task with sleep 300
> > kill the task
> > sleep 300 is still running
> 
> Hi, @zhongjiajie @Radeity @SbloodyS Could you guys please take a time to see if this bug can be reproduced?
> 
> On my mac, the task is killed at 14:48 image
> 
> but `sleep 300` is still running at 14:49 截屏2023-07-13 14 50 13

Hi, @rickchengx , I can reproduce this bug on my linux server.

**Comment by Radeity** [comment]
> I hope we can remove the `doKill` method, the kill operation should be implemented by task plugin.

Agree with you, current kill operation seems mess. 

However, for this PR, this fix does work. If `process.destroy` cannot kill the subprocess, we can only get subprocess id in  `getPidsStr` before the main process is killed.

**Comment by rickchengx** [comment]
@ruanwenjun I think we could fix this bug first through this PR (If the child process cannot be stopped, I think the impact of this bug is relatively large). And we could improve the kill logic to use the `cancelApplication()` in each task plugin later. It looks like a lot of work. WDYT

cc @Radeity @zhongjiajie @SbloodyS

**Review by Radeity — APPROVED** [review]
+1

**Review by SbloodyS — APPROVED** [review]
+1

**Pull request merged**